### ISSUE TITLE
Reemulate from save state

### ISSUE THREAD

**Issue by thrust26** [stella-emu/stella#1052]
When a trap or breakpoint is hit, it would be cool if one could emulate from the last save state including any user inputs. So any user inputs would have to recorded too.

**Comment by JetSetIlly** [comment]
@thrust26 With regard to the specific problem that prompted this, there might be a relatively simple solution.

The problem: If user-input has indirectly caused the emulation to be halted there can be a problem stepping back to the previous time machine state and then replaying the correct code path. Any user input that has happened since the previous saved state has been lost.

A simple solution that might be sufficient would be to make a time machine save state on every user input in addition to the normal interval. This would have a similar effect I believe.

For example:
```
Frame n ->   n+1 ->   n+2 ->   n+3 ->   n+4 ->  n+ 5 ->  n+6 >   input   >  break into debugger
```

Currently, stepping back from the breakpoint goes to "n+6". From here it's impossible to step forward and recreate the initial debugger state (because the input state has been lost). However, if the debugger steps back to the "input" state then it would be possible.

It would be tricky to get right but it might be easier than recording user input events and reinserting into the input stream.

If implementing this solution then care should be taken over what happens when the user steps-back further than the input state.

**Comment by thrust26** [comment]
For e.g. joystick input, that's a good idea. However for paddle (and trakball) games, this would create a **lot** of save states. Or do I miss something here?

**Comment by JetSetIlly** [comment]
Yes. Paddle input would create a lot of saves. For the specific problem wanting to be solved, keeping only the most recent input saved state would work I think.
